### DISCUSSION TITLE
🐕 Huskyとは？

### DISCUSSION THREAD

**Discussion by kagomen** [lgtm-factory/lgtm-factory#57] [Scrap]
イシュー化するにあたって、以下を読む

~~- https://qiita.com/mu-suke08/items/43a492fda5cd71a31506~~ 【追記】書き方古いので注意！
- https://zenn.dev/risu729/articles/latest-husky-lint-staged

**Comment by kagomen** [comment]
> レビューをもらう前に必ず行う、testだったりlintだったりを代わりにやってくれるのがhuskyです。
当初はこの部分をGithub Actionsで自動化していたのですが、エラーを吐いた場合のコミットし直しだったり、CIの結果を待つ時間自体が非効率であったためhuskyを使うようになりました。

Github Actionsでは足りない、と！

**Reply by kagomen** [reply]
> lint-stagedとは、Gitでステージ上の、つまり変更されたファイルに対してのみlintやformatをするためのnpmライブラリです。

**Reply by kagomen** [reply]
ファスコンでもコミット前に `npm run format` の実行を忘れたことがあったと思う
ので、やっぱり導入したい・・・

**Reply by ghost** [reply]
逆にPrettierいれててもnpm run format実行しないと
意味ない...?のね...

**Reply by kagomen** [reply]
@soraOrange 

ですです
そんなの忘れちゃいますよね😂

**Reply by ghost** [reply]
@kagomen めっちゃ勉強なりました.....

**Comment by kagomen** [comment]
> 2021/3/26のhusky v5で、このように package.json で設定する方法はサポートされなくなりました。
どういうわけか2022年の記事にも書かれていたりしましたが、これは(v5以上を使っている限り)無意味です。

1つめのQiitaの記事の更新日は2024年になってたけど↑の書き方だったので注意

**Comment by kagomen** [comment]
> Git フック とは、Gitで特定のコマンドの実行前後にスクリプトを実行するための仕組みです。

**Reply by kagomen** [reply]
.git/hooksの中にGitHookの設定ファイルがある

<img width="240" alt="スクリーンショット 2024-08-11 22 47 13" src="https://github.com/user-attachments/assets/041d7c88-edd9-4cc3-9f82-af3d1ef0b2b2">

<img width="292" alt="スクリーンショット 2024-08-11 22 47 22" src="https://github.com/user-attachments/assets/a2863517-9812-425f-9102-b6f09e98b936">

**Reply by kagomen** [reply]
ファイルの中身こんなん
これをいじったらGitコマンド用のスクリプトを設定できるけど、何書いてあるのか全然わからない、無理💧
これを使いやすくしてくれてるのが、Husky！🐕
<img width="622" alt="スクリーンショット 2024-08-11 22 48 51" src="https://github.com/user-attachments/assets/af035e4e-a9e4-4d57-bfea-e3a786e70c14">

**Comment by kagomen** [comment]
> huskyとは、Gitフックで任意のプログラムを実行するためのnpmライブラリです。

**Reply by kagomen** [reply]
husky initしたら.husky/pre-commit が作成される
pre-commit というのは、GitHooksの一種で、コミットの前に実行するコマンドを設定できる！

pre-commit の他にも.git/hooksフォルダに色んな設定ファイルあったけど、希望のhooksに応じて.huskyディレクトリにファイル作ればいい！

**Comment by kagomen** [comment]
> lint-stagedとは、Gitでステージ上の、つまり変更されたファイルに対してのみlintやformatをするためのnpmライブラリです。

**Reply by kagomen** [reply]
git add したものを対象に、lintやformatをかけるライブラリ、なので、それ以外のものは対象外
なので、他の人が編集したソースコードなんかには、フォーマットが適用されない

ということは、prettierとhusky, lint-stagedは早々に入れておいた方が良さそう？

**Reply by kagomen** [reply]
Huskyだけでprettierを実行させるのと、huskyにlint-stagedも加えてprettierを実行させるのと、何が違う？どんなメリットがある？

---

HuskyだけでPrettierを実行する場合と、Husky + lint-stagedでPrettierを実行する場合の主な違いとメリットは以下の通りです：

1. Huskyのみを使用する場合:

メリット:
- セットアップが比較的簡単
- すべてのファイルに対して常にPrettierが実行される

デメリット:
- コミット時に全ファイルに対してPrettierが実行されるため、時間がかかる
- 変更していないファイルも含めて全てのファイルがフォーマットされる

2. Husky + lint-stagedを使用する場合:

メリット:
- ステージングされた（コミット予定の）ファイルのみにPrettierが実行される
- コミット時の処理が高速化される（変更ファイルのみを対象とするため）
- 不要なファイルの変更を避けられる
- 複数のlintツールを組み合わせて使用しやすい

デメリット:
- セットアップがやや複雑になる

lint-stagedを追加することの主なメリット:

1. パフォーマンスの向上:
   変更されたファイルのみを処理するため、大規模プロジェクトでも高速に動作します。

2. 柔軟性:
   ファイルタイプごとに異なるlintツールを適用できます。例えば、.jsファイルにはESLint、.cssファイルにはStylelintを使用するといった設定が可能です。

3. 予期しない変更の防止:
   変更していないファイルが意図せずフォーマットされることを防ぎます。

4. 段階的な導入:
   既存のプロジェクトに段階的にlintツールを導入する際に便利です。

5. チーム開発の効率化:
   コミット前に自動的にコードスタイルを統一することで、レビューの効率が上がります。

結論として、Husky + lint-stagedの組み合わせは、特に中〜大規模のプロジェクトや、複数の開発者が関わるプロジェクトで非常に有効です。小規模なプロジェクトや個人開発の場合は、Huskyのみでも十分な場合があります。

**Reply by kagomen** [reply]
huskyとlint-stagedを導入したあとに、既存のソースコードもフォーマットかけたい場合はどうしたらいい？

---

huskyとlint-stagedを導入した後に、既存のソースコード全体にフォーマットをかけたい場合、以下の方法があります：

1. NPM スクリプトを使用する方法:

package.json に以下のようなスクリプトを追加します：

```json
{
  "scripts": {
    "format": "prettier --write ."
  }
}
```

そして、以下のコマンドを実行します：

```
npm run format
```

2. Prettier CLI を直接使用する方法:

プロジェクトのルートディレクトリで以下のコマンドを実行します：

```
npx prettier --write .
```

3. 特定のディレクトリやファイルタイプを指定する方法:

特定のディレクトリやファイルタイプのみをフォーマットしたい場合は、以下のようにコマンドを調整できます：

```
npx prettier --write "src/**/*.{js,jsx,ts,tsx,css,scss}"
```

4. VS Code の Prettier 拡張機能を使用する方法:

VS Code で Prettier 拡張機能をインストールし、「Format Document」または「Format Folder」コマンドを使用してフォーマットを適用します。

5. Git hooks を一時的に無効化する方法:

既存のコードをフォーマットする際に Git hooks が邪魔になる場合は、以下のコマンドで一時的に無効化できます：

```
HUSKY=0 npx prettier --write .
```

注意点:

- 大規模な変更を行う前に、必ずバージョン管理システム（Git など）でコミットしておくことをお勧めします。
- チームで開発している場合は、一斉にフォーマットを適用するタイミングをチームメンバーと調整することが重要です。
- フォーマット後は、アプリケーションが正常に動作することを確認してください。

これらの方法を使用することで、既存のソースコード全体に Prettier によるフォーマットを適用できます。プロジェクトの規模や要件に応じて最適な方法を選択してください。

**Reply by kagomen** [reply]
たしかに

じゃあ、

> prettierとhusky, lint-stagedは早々に入れておいた方が良さそう？

に関して、特に急ぐ必要はない！あとからでも全然OK

**Reply by kagomen** [reply]
package.jsonと.lintstagedrc、どちらでも設定できるみたい
どっちが良いんだろう？

---

lint-stagedの設定を行う場合、package.jsonと.lintstagedrcのどちらを使用するかは、プロジェクトの要件や個人の好みによって異なります。両方にメリットとデメリットがありますので、それぞれを比較してみましょう。

1. package.jsonに設定する場合

メリット:
- 設定ファイルの数が少なくなる（1つのファイルにまとまる）
- NPMスクリプトと lint-staged の設定が近くにあるため、関連性が分かりやすい

デメリット:
- package.jsonが大きくなりすぎる可能性がある
- lint-stagedの設定が複雑な場合、package.jsonの可読性が低下する可能性がある

2. .lintstagedrcを使用する場合

メリット:
- lint-stagedの設定が独立しているため、管理しやすい
- 複雑な設定の場合、より読みやすく保守しやすい
- JSONだけでなく、JavaScript形式でも書けるため、より柔軟な設定が可能

デメリット:
- プロジェクトのルートディレクトリにファイルが1つ増える

どちらを選択するかの指針:

1. 小規模なプロジェクトや、シンプルな lint-staged の設定の場合:
   - package.json に設定を含める方が良いでしょう。

2. 大規模なプロジェクトや、複雑な lint-staged の設定の場合:
   - .lintstagedrc（または .lintstagedrc.json, .lintstagedrc.js）を使用する方が良いでしょう。

3. プロジェクトの一貫性:
   - すでに他の設定（例：.eslintrc, .prettierrc）を独立したファイルで管理している場合は、一貫性のために.lintstagedrcを使用するのが良いでしょう。

4. チームの好み:
   - チームで開発している場合は、チームメンバーの意見も考慮に入れてください。

5. 将来の拡張性:
   - 将来的に lint-staged の設定が複雑になる可能性がある場合は、最初から .lintstagedrc を使用することを検討してください。

個人的な推奨:
一般的には、設定の管理のしやすさと将来の拡張性を考慮して、.lintstagedrcを使用することをお勧めします。特に、他の設定ファイル（.eslintrc, .prettierrcなど）も独立したファイルで管理している場合は、一貫性を保つために.lintstagedrcを使用するのが良いでしょう。

ただし、最終的にはプロジェクトの要件やチームの好みに応じて判断してください。どちらの方法を選んでも、lint-stagedの機能性には影響ありません。

**Reply by kagomen** [reply]
一貫性を保つ、というなら、create-next-appで自動作成される .eslintrc.jsonに合わせるのが良さそう？
.prettierrc.json, .lintstagedrc.jsonで揃える・・・？

実装者に任せる

**Reply by kagomen** [reply]
.XXXrcファイルと.XXXrc.jsonファイルの主な違い、およびそれぞれのメリットとデメリットについて説明します。

主な違い:
1. ファイル拡張子: .XXXrcは拡張子なし、.XXXrc.jsonは.json拡張子あり
2. パース方法: .XXXrcは複数のフォーマットをサポート可能、.XXXrc.jsonは常にJSONとしてパース

.XXXrcファイル:

メリット:
1. 柔軟性: JSON, YAML, または独自フォーマットなど、複数のフォーマットをサポートできる
2. 簡潔性: 拡張子がないためファイル名が短くなる
3. 伝統的: 多くのUNIXツールで使用されている慣習的な命名規則

デメリット:
1. あいまいさ: ファイル形式が明示的でないため、内容を見るまでフォーマットが分からない
2. エディタサポート: 一部のエディタで構文ハイライトやバリデーションが効きにくい場合がある
3. パース時のオーバーヘッド: 形式を判断してからパースする必要があるため、若干のパフォーマンス低下の可能性

.XXXrc.jsonファイル:

メリット:
1. 明確性: ファイル形式がJSONであることが明確
2. エディタサポート: 多くのエディタでJSONファイルとして認識され、構文ハイライトやバリデーションが効く
3. パース効率: 常にJSONとしてパースされるため、形式判断のオーバーヘッドがない
4. スキーマ検証: JSONスキーマを使用した検証が容易

デメリット:
1. 冗長性: ファイル名が長くなる
2. 柔軟性の欠如: JSON形式のみをサポート
3. コメント非対応: 標準のJSONはコメントをサポートしていない

選択の指針:

1. .XXXrcを選ぶ場合:
   - 複数のフォーマットをサポートしたい場合
   - 短いファイル名を好む場合
   - 伝統的なUNIX風の命名規則を好む場合

2. .XXXrc.jsonを選ぶ場合:
   - 明示的にJSON形式であることを示したい場合
   - エディタの支援機能（構文ハイライト、バリデーションなど）を最大限活用したい場合
   - JSONスキーマを使用した検証を行いたい場合
   - パース効率を重視する場合

実際の選択は、プロジェクトの要件、チームの好み、使用するツールやエディタとの相性などを考慮して決定するのが良いでしょう。多くの現代的なプロジェクトでは、明示性と開発者体験の向上のため、.XXXrc.json形式を選択する傾向にあります。

**Reply by kagomen** [reply]
"rc" は "run commands" または "runtime configuration" の略称です。この用語には長い歴史があり、UNIXシステムの初期から使用されてきました。

rcの起源と意味:

1. 歴史的な起源:
   - "rc" の使用は1965年の MIT の CTSS (Compatible Time-Sharing System) オペレーティングシステムにまで遡ります。
   - "run commands" の略として使用され始めました。

2. 現代的な解釈:
   - "runtime configuration" としても解釈されることが多くなっています。
   - プログラムやシステムの起動時に読み込まれる設定ファイルを指す一般的な用語となっています。

3. 一般的な使用:
   - .bashrc, .vimrc, .npmrc など、多くのプログラムやツールの設定ファイルで使用されています。
   - これらのファイルは通常、プログラムの起動時に読み込まれ、動作をカスタマイズするために使用されます。

4. 命名規則:
   - 通常、プログラム名または関連する概念の後に "rc" を付けて命名されます（例: .eslintrc, .babelrc）。
   - ドットファイル（.で始まるファイル）として実装されることが多く、UNIXシステムでは隠しファイルとして扱われます。

5. 現代的な開発環境での使用:
   - 多くの開発ツールやフレームワークが、この命名規則を採用して設定ファイルを実装しています。
   - JSONやYAMLなど、より構造化されたフォーマットを使用することも増えていますが、"rc" という名称は依然として広く使用されています。

rcファイルの特徴:

- プログラムの起動時に自動的に読み込まれる
- ユーザーやプロジェクト固有の設定を定義する
- しばしばホームディレクトリやプロジェクトのルートディレクトリに配置される
- テキストベースで編集可能

まとめると、"rc" は設定ファイルを示す一般的な規則となっており、プログラムの動作をカスタマイズするための重要な役割を果たしています。現代のソフトウェア開発においても、この命名規則は広く採用され続けています。

**Comment by kagomen** [comment]
`prettier --write .`、`"**/*": "prettier --write --ignore-unknown"`、`"*": "prettier --write"` の3つの方法には、それぞれ特徴と違いがあります。以下に詳細を説明します：

1. `prettier --write .`

   - これはコマンドラインで直接実行する形式です。
   - カレントディレクトリ（`.`）とそのすべてのサブディレクトリ内の、Prettierがサポートするファイルタイプに対して適用されます。
   - デフォルトで `--ignore-unknown` の動作をします（Prettierが認識しないファイルタイプは無視します）。
   - `.prettierignore` ファイルの設定を尊重します。

2. `"**/*": "prettier --write --ignore-unknown"`

   - これは通常、lint-staged や他のツールの設定で使用されます。
   - すべてのディレクトリとサブディレクトリを含む全ファイルにマッチします。
   - `--ignore-unknown` オプションにより、Prettierが認識しないファイルタイプを明示的にスキップします。
   - lint-staged を使用している場合、ステージングされたファイルのみに適用されます。

3. `"*": "prettier --write"`

   - これも通常、lint-staged や他のツールの設定で使用されます。
   - カレントディレクトリ内のファイルのみにマッチし、サブディレクトリは含みません。
   - `--ignore-unknown` オプションがないため、Prettierが認識しないファイルタイプに対しても処理を試みる可能性があります。
   - lint-staged を使用している場合、ステージングされたファイルのみに適用されます。

主な違いと注意点：

- 適用範囲：`prettier --write .` と `"**/*"` はサブディレクトリを含みますが、`"*"` は現在のディレクトリのみです。
- 安全性：`prettier --write .` と `"**/*": "prettier --write --ignore-unknown"` は未知のファイルタイプを安全に無視しますが、`"*": "prettier --write"` はそうではありません。
- 使用コンテキスト：`prettier --write .` は直接のコマンド実行に適していますが、他の2つは通常ツールの設定内で使用されます。
- ファイル選択：lint-staged を使用する場合、後者2つの方法はステージングされたファイルのみに適用されますが、`prettier --write .` はすべての該当ファイルに適用されます。

実際の使用では、コンテキストと目的に応じて適切な方法を選択することが重要です。多くの場合、`prettier --write .` や `"**/*": "prettier --write --ignore-unknown"` が安全で包括的なオプションとなります。

**Reply by kagomen** [reply]
サブディレクトリ（subdirectory）とは、あるディレクトリ（フォルダ）の中に存在する別のディレクトリのことを指します。つまり、ディレクトリの階層構造における「子」のディレクトリです。

例えば、以下のようなディレクトリ構造があるとします：

```
project/
│
├── src/
│   ├── components/
│   │   └── Button.js
│   └── pages/
│       └── Home.js
│
├── public/
│   └── index.html
│
└── package.json
```

この構造では：

- `project` がルートディレクトリ（親ディレクトリ）です。
- `src` と `public` は `project` のサブディレクトリです。
- `components` と `pages` は `src` のサブディレクトリです。

サブディレクトリの重要性：

1. 組織化：大規模なプロジェクトでファイルを論理的に整理するのに役立ちます。

2. モジュール性：関連するファイルをグループ化し、コードの構造を明確にします。

3. パス指定：ファイルやリソースへのパスを指定する際に使用されます。

4. アクセス制御：特定のディレクトリやサブディレクトリに対してアクセス権限を設定できます。

5. ビルドプロセス：多くのビルドツールやタスクランナーは、特定のディレクトリ構造を前提としています。

Prettierやlint-stagedの設定において、サブディレクトリの考慮は重要です。例えば：

- `"*": "prettier --write"` はルートディレクトリの直下のファイルのみを対象とします。
- `"**/*": "prettier --write"` はすべてのサブディレクトリを含む全ファイルを対象とします。

プロジェクトの構造に応じて適切な設定を選択することで、意図した範囲のファイルに対して正しくツールを適用できます。

**Reply by kagomen** [reply]
--ignore-unknown オプションと .prettierignore ファイル（または .prettierrc の ignore 設定）は異なる目的を持っており、独立して動作します。

1. --ignore-unknown オプション:
   - このオプションは、Prettier が認識しないファイルタイプ（サポートしていない拡張子を持つファイル）を自動的にスキップします。
   - 例えば、.jpg や .png などの画像ファイル、あるいは Prettier がサポートしていない特殊なテキストファイルフォーマットなどが該当します。

2. .prettierignore ファイル（または .prettierrc の ignore 設定）:
   - これらは、Prettier がサポートするファイルタイプであっても、明示的に無視したいファイルやディレクトリを指定するためのものです。
   - 例えば、生成されたコードや、特定のライブラリファイルなどを無視するために使用します。

重要なポイント:

- --ignore-unknown オプションは、.prettierignore や .prettierrc の ignore 設定とは独立して動作します。
- .prettierignore で指定されたファイルは、--ignore-unknown オプションの有無にかかわらず常に無視されます。
- --ignore-unknown を使用しても、.prettierignore の設定は引き続き尊重されます。

つまり、--ignore-unknown オプションを使用しても、.prettierignore で設定したファイルは別途スキップされます。このオプションは単に、Prettier が処理できないファイルタイプを自動的にスキップする機能を提供するだけです。

例えば:

1. .prettierignore で node_modules ディレクトリを無視するよう設定している場合、--ignore-unknown の有無にかかわらず、node_modules 内のファイルは処理されません。

2. --ignore-unknown を使用すると、.jpg ファイルなどの Prettier が処理できないファイルは自動的にスキップされますが、これは .prettierignore の設定とは無関係です。

したがって、両方の機能を併用することで、より細かく制御された Prettier の適用が可能になります。不要なファイルの処理を避けつつ、意図しないファイルタイプへの適用も防ぐことができます。
